Spec draft for transient input hit test

bialpio opened the pull request:
This PR brings in spec text for hit test for transient input sources.


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/immersive-web/hit-test/pull/64.html" title="Last updated on Dec 28, 2019, 12:13 AM UTC (0ee0fd6)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/immersive-web/hit-test/64/f3db0c6...0ee0fd6.html" title="Last updated on Dec 28, 2019, 12:13 AM UTC (0ee0fd6)">Diff</a>

toji commented on index.bs:
```diff
@@ -229,85 +303,176 @@ The {{XRHitTestResult/getPose(baseSpace)}} method, when invoked on {{XRHitTestRe
 
 </div>
 
+<div class="algorithm" data-algorithm="compute-hit-test-results-for-transient-input">
```
Related to the above point of some transient input sources only existing for the space of a select event, I want to clarify the intended "lifetime" of a hit test result. This algorithm suggests that hit test results can only be returned while the input source is in the active list, right? The bit about animation frames above made me wonder if the intent was to return all hit test results that occurred since the previous animation frame, regardless of if the input source was still available.

bialpio replied:
> This algorithm suggests that hit test results can only be returned while the input source is in the active list, right?

Correct. Chrome's case is a bit specific because as far as I can tell, our transient input sources always are alive for at least one animation frame (so results are available both for the select events, and during rAF callbacks), but the intent here is for the API to not hand out objects that are already in zombie state (transient hit test result contains XRInputSource so I don't think we want to deliver results whose the input source is no longer active).

bialpio replied:
To clarify, the transient input sources in AR case are alive for at least one animation frame.

toji commented on index.bs:
```diff
@@ -133,12 +137,35 @@ The <dfn dict-member for="XRHitTestOptionsInit">space</dfn> dictionary member sp
+
+<script type="idl">
+dictionary XRTransientInputHitTestOptionsInit {
+  required DOMString profile;
```
I know this has the "transient" name plastered all over it, but I still suspect some people will see this and assume they can use it as an easy way to subscribe to hit tests for all inputSources with the given profile, transient or otherwise. And, honestly, maybe that's not a terrible idea? Haven't really thought through the implementation details, though.

bialpio replied:
Current wording of the spec blocks this use case (for transient hit test results to be computed, the input source needs to be categorized as transient). I'd say we should avoid having 2 ways of doing the same thing and this variant of the API should exist _only_ to enable hit testing for transient input sources, since this is the only case which the regular hit test cannot solve.

bialpio commented on index.bs:
```diff
@@ -229,85 +303,176 @@ The {{XRHitTestResult/getPose(baseSpace)}} method, when invoked on {{XRHitTestRe
 ==========================
 
-Every time the user agent processes [=/XR animation frame=] for {{XRSession}}, it needs to <dfn>compute hit test results</dfn>. This MUST happen prior to invoking any of the callbacks from [=/list of animation frame callbacks=] of a given {{XRSession}}.
+Every time the user agent processes [=/XR animation frame=] for {{XRSession}}, it needs to <dfn>compute all hit test results</dfn>. This MUST happen prior to invoking any of the callbacks from [=/list of animation frame callbacks=] of a given {{XRSession}}.
```
I'll remove this phrasing, the algorithm should be clear that it happens for both animation frames and select events. When we create hit test source, we'll add new algorithm to session's list of frame updates and then apply frame updates both for animation frame and for frames passed in to select events.

https://immersive-web.github.io/webxr/#xrframe-apply-frame-updates

bialpio replied:
I'll remove this phrasing, the algorithm should be clear that it happens for both animation frames and select events. When we create hit test source, we'll add new algorithm to session's list of frame updates and then apply frame updates both for animation frame and for frames passed in to select events.

bialpio replied:
Removed, resolving the conversation.

bialpio commented on index.bs:
```diff
@@ -133,12 +137,35 @@ The <dfn dict-member for="XRHitTestOptionsInit">space</dfn> dictionary member sp
+};
+</script>
+
+The <dfn dict-member for="XRTransientInputHitTestOptionsInit">profile</dfn> dictionary member specifies an [=XRInputSource/input profile name=] of the transient input source that will be used to compute hit test results.
```
Fyi, I'm currently using `generic-touchscreen` in my prototype implementation.

bialpio replied:
Created #69 to discuss.

toji commented:
What's the intent behind merging this now? It hasn't gotten approval from anyone else, and my understanding was that a different variant of the API was going to be proposed?

bialpio commented:
There was no yet-to-be-addressed feedback on this PR for a while and the currently open issues did not warrant an API redesign (they are not specific to transient input sources), so I decided to merge it.
